<a href="https://colab.research.google.com/github/adhang/data-science-digitalskola/blob/update/07.%20Intermediate%20Pandas%20Dataframe/Learn%20-%20Intermediate%20Pandas%20Dataframe.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Intermediate Pandas Dataframe
Author: Adhang Muntaha Muhammad

[![LinkedIn](https://img.shields.io/badge/linkedin-0077B5?style=for-the-badge&logo=linkedin&logoColor=white&link=https://www.linkedin.com/in/adhangmuntaha/)](https://www.linkedin.com/in/adhangmuntaha/)
[![GitHub](https://img.shields.io/badge/github-121011?style=for-the-badge&logo=github&logoColor=white&link=https://github.com/adhang)](https://github.com/adhang)
[![Kaggle](https://img.shields.io/badge/kaggle-20BEFF?style=for-the-badge&logo=kaggle&logoColor=white&link=https://www.kaggle.com/adhang)](https://www.kaggle.com/adhang)
[![Tableau](https://img.shields.io/badge/tableau-E97627?style=for-the-badge&logo=tableau&logoColor=white&link=https://public.tableau.com/app/profile/adhang)](https://public.tableau.com/app/profile/adhang)
___
**Contents**
- Importing Libraries
- Reading Dataset
- Sorting Dataframe
- Filtering Dataframe
- Creating New Column
- Grouping Dataframe
- Merging Dataframe

**Important Notes**<br>
If you re-run this notebook, maybe the dataframe won't be the same, since I was using random sampling to get the data. So, some of my descriptions of the results may not be suitable to your results. You will find out later when you read the entire contents of this notebook.

# Importing Libraries

In [ ]:
import pandas as pd

# Reading Dataset
For this notebook, I will use my mentor's dataset from GitHub. You can check his works [here](https://github.com/densaiko).

In [ ]:
file_path = 'https://raw.githubusercontent.com/densaiko/data_science_learning/main/dataset/Human%20Capital.csv'

data = pd.read_csv(file_path)
data.head()

employee_id         department  ... avg_training_score is_promoted
0        65438  Sales & Marketing  ...               49.0           0
1        65141         Operations  ...               60.0           0
2         7513  Sales & Marketing  ...               50.0           0
3         2542  Sales & Marketing  ...               50.0           0
4        48945         Technology  ...               73.0           0

[5 rows x 13 columns]

Get 10 samples, random from the dataset.

In [ ]:
data_10 = data.sample(n=10)
data_10

employee_id         department  ... avg_training_score is_promoted
50374        15356         Technology  ...               76.0           0
22091        34094         Technology  ...               78.0           0
10083         2328        Procurement  ...               69.0           0
11436        46485          Analytics  ...               87.0           0
36604         6585  Sales & Marketing  ...               51.0           0
26905         4518  Sales & Marketing  ...               48.0           0
11793        77474          Analytics  ...               86.0           0
835          57617        Procurement  ...               70.0           0
21318        69856        Procurement  ...               70.0           0
14320        28207  Sales & Marketing  ...               52.0           0

[10 rows x 13 columns]

# Sorting Dataframe

## Sort Dataframe by Specific Column
Syntax
```
dataframe.sort_values(by='column_name')
```

By default, sorting will be done in ascending order.

In [ ]:
# sort dataframe by employee_id column
data_10.sort_values(by='employee_id')

employee_id         department  ... avg_training_score is_promoted
10083         2328        Procurement  ...               69.0           0
26905         4518  Sales & Marketing  ...               48.0           0
36604         6585  Sales & Marketing  ...               51.0           0
50374        15356         Technology  ...               76.0           0
14320        28207  Sales & Marketing  ...               52.0           0
22091        34094         Technology  ...               78.0           0
11436        46485          Analytics  ...               87.0           0
835          57617        Procurement  ...               70.0           0
21318        69856        Procurement  ...               70.0           0
11793        77474          Analytics  ...               86.0           0

[10 rows x 13 columns]

We can change the sorting order by setting the `ascending` parameter.

In [ ]:
# sort in descending order
data_10.sort_values(by='employee_id', ascending=False)

employee_id         department  ... avg_training_score is_promoted
11793        77474          Analytics  ...               86.0           0
21318        69856        Procurement  ...               70.0           0
835          57617        Procurement  ...               70.0           0
11436        46485          Analytics  ...               87.0           0
22091        34094         Technology  ...               78.0           0
14320        28207  Sales & Marketing  ...               52.0           0
50374        15356         Technology  ...               76.0           0
36604         6585  Sales & Marketing  ...               51.0           0
26905         4518  Sales & Marketing  ...               48.0           0
10083         2328        Procurement  ...               69.0           0

[10 rows x 13 columns]

## Sort Single Column
We can take a specific column and then sort it. The outpus will be a series.

In [ ]:
# get employee_id column and sort it
data_10['employee_id'].sort_values()

10083     2328
26905     4518
36604     6585
50374    15356
14320    28207
22091    34094
11436    46485
835      57617
21318    69856
11793    77474
Name: employee_id, dtype: int64

In [ ]:
# sort in descending order
data_10['employee_id'].sort_values(ascending=False)

11793    77474
21318    69856
835      57617
11436    46485
22091    34094
14320    28207
50374    15356
36604     6585
26905     4518
10083     2328
Name: employee_id, dtype: int64

## Sort by Multiple Columns
Sort by first column, then the second column.

In [ ]:
data_10.sort_values(by=['department','age'])

employee_id         department  ... avg_training_score is_promoted
11793        77474          Analytics  ...               86.0           0
11436        46485          Analytics  ...               87.0           0
835          57617        Procurement  ...               70.0           0
21318        69856        Procurement  ...               70.0           0
10083         2328        Procurement  ...               69.0           0
36604         6585  Sales & Marketing  ...               51.0           0
14320        28207  Sales & Marketing  ...               52.0           0
26905         4518  Sales & Marketing  ...               48.0           0
22091        34094         Technology  ...               78.0           0
50374        15356         Technology  ...               76.0           0

[10 rows x 13 columns]

Sort the data by department, then we sort by age. You can see at the first 2 rows have the same `department` name. Therefore, the order of those first 2 rows is determined by the `age` column.

# Filtering Dataframe
Let's create a new dataframe using random sampling.

In [ ]:
data_20 = data.sample(n=20)
data_20.shape

(20, 13)

## Get Specific Columns
Getting some specific columns in Pandas is similar to the `SELECT` statement in SQL. For example, we want to get `department` and `awards_won` columns.

### Using Square Brackets
Use double square brackets `[[]]` to create a dataframe. If we only use a single bracket `[]`, it will create a series. It's a native way to create a series or a dataframe.

In [ ]:
data_20[['department','awards_won']].head()

department  awards_won
29967         Technology           0
17921            Finance           0
46538                 HR           0
44257  Sales & Marketing           0
54629         Operations           0

### Using Function
Pandas has a built-in method to select specific columns.

In [ ]:
data_20.filter(items=['department','awards_won']).head()

department  awards_won
29967         Technology           0
17921            Finance           0
46538                 HR           0
44257  Sales & Marketing           0
54629         Operations           0

## Data Selection
- `loc` - used to select rows or columns by its name
- `iloc` - used to select rows or columns by its index

I can say, `loc` stands for `location`, and `iloc` stands for `integer location` or `index location`. Just to make it easier to memorize the difference.

But first, I'm going to reset the index of my dataframe.

In [ ]:
data_20 = data.sample(n=20)
data_20.reset_index(inplace=True)
data_20.head()

index  employee_id  ... avg_training_score is_promoted
0  53850        61507  ...               48.0           0
1  20985        32689  ...               49.0           0
2   8827        10569  ...               58.0           0
3  51182        27042  ...               46.0           0
4  49323        14123  ...               64.0           0

[5 rows x 14 columns]

You can see that the index from the dataframe becomes a new column. We don't want this. So, We have to use the `drop` parameter when resetting the index.

In [ ]:
data_20 = data.sample(n=20)
data_20.reset_index(inplace=True, drop=True)
data_20.head()

employee_id         department  ... avg_training_score is_promoted
0        59352  Sales & Marketing  ...               53.0           0
1          133         Technology  ...               77.0           0
2        15194          Analytics  ...               88.0           0
3        45914  Sales & Marketing  ...               43.0           1
4        21728  Sales & Marketing  ...               54.0           0

[5 rows x 13 columns]

### loc
Let's say, we want to get the first 5 rows for the `is_promoted` and `gender` columns. Since we will use column names, we have to use the `loc` method.

In [ ]:
data_20.loc[3:5, ['is_promoted', 'gender']]

is_promoted gender
3            0      f
4            1      m
5            0      f

### iloc
Let's see the last 10 rows using `tail()`.

In [ ]:
data_20.tail(10)

employee_id         department  ... avg_training_score is_promoted
10        16316         Technology  ...               78.0           0
11        28606         Operations  ...               57.0           0
12         2220  Sales & Marketing  ...               48.0           0
13        63460          Analytics  ...               83.0           0
14        63247         Technology  ...               82.0           0
15         2085  Sales & Marketing  ...               50.0           1
16         6853          Analytics  ...               82.0           0
17         9930          Analytics  ...               86.0           0
18        58557         Technology  ...               80.0           0
19        61727            Finance  ...                NaN           0

[10 rows x 13 columns]

For this case, we want to get the last 10 rows for the `recruitment_channel` and `no_of_trainings` columns. But this time, we will use the index values.

Just like a Numpy array, the index of rows is started from 0, and the index of columns is started from 0. So, if we can say that `recruitment_channel` is located in the 6th column (index = 5) and `no_of_trainings` column is located in the 7th column (index = 6). Remember, the upper-bound of slicing is exclusive (not included).

In [ ]:
data_20.iloc[-10:, 5:7]

recruitment_channel  no_of_trainings
10            sourcing                1
11            sourcing                1
12            sourcing                2
13               other                2
14               other                1
15               other                1
16               other                2
17            sourcing                1
18               other                1
19               other                1

### Select By Conditions
What if we want to filter our data using a condition? Let's say we want to get data where the `age` is above 35.

In [ ]:
filter_data = data_20.loc[data_20['age'] > 35]
filter_data

employee_id         department  ... avg_training_score is_promoted
1           133         Technology  ...               77.0           0
3         45914  Sales & Marketing  ...               43.0           1
7         13501  Sales & Marketing  ...               52.0           0
9         65977          Analytics  ...               86.0           0
11         1583        Procurement  ...               73.0           0
16        52613                R&D  ...               83.0           0

[6 rows x 13 columns]

As you can see, we get new dataframe where the `age` is above 35. How about using multiple conditions? For example, we want to get data where the `age` is above 35 and the `length_of_service` is above 4.

In [ ]:
filter_data = data_20.loc[(data_20['age'] > 35) & (data_20['length_of_service'] > 4)]
filter_data

employee_id         department  ... avg_training_score is_promoted
1          133         Technology  ...               77.0           0
3        45914  Sales & Marketing  ...               43.0           1
7        13501  Sales & Marketing  ...               52.0           0

[3 rows x 13 columns]

We have filtered our data using 2 conditions. Filtering data using conditions in Pandas is similar to the `WHERE` statement in SQL.

Then, how about selecting specific columns and filtering the data at the same time?

This time, I want to get data where the `age` is above 35 and the `lenght_of_service` is above 4. Beside that, I only want to display `employee_id`, `department`, `region`, `age`, and `length_of_service` columns.

In [ ]:
filter_data = data_20.loc[(data_20['age'] > 35) & (data_20['length_of_service'] > 4), ['employee_id','department','region','age','length_of_service']]
filter_data

employee_id         department     region  age  length_of_service
1          133         Technology  region_15   46                  6
3        45914  Sales & Marketing   region_2   36                  9
7        13501  Sales & Marketing  region_26   41                  7

The code below is another way to do the same thing as the above code.

In [ ]:
filter_data = data_20[['employee_id','department','region','age','length_of_service']].loc[(data_20['age'] > 35) & (data_20['length_of_service'] > 4)]
filter_data

employee_id         department     region  age  length_of_service
1          133         Technology  region_15   46                  6
3        45914  Sales & Marketing   region_2   36                  9
7        13501  Sales & Marketing  region_26   41                  7

# Creating New Column
Sometimes, we need to create some new columns. For example, we have a data like this:

| **id** | **name** | **quiz** | **homework** |
|:------:|:--------:|:--------:|:------------:|
|    1   |  Adhang  |    80    |      85      |
|    2   |  Muntaha |    90    |      85      |
|    3   | Muhammad |    85    |      90      |

We have `quiz` and `homework` scores. Then, we need to create a new column to store the `total` score. We can do it using Pandas.

For this dataset, I will use a dummy calculation that doesn't have any meaning (or usage).

In [ ]:
# sum 'age' and 'length_of_service'
filter_data['dummy_sum'] = filter_data['age'] + filter_data['length_of_service']
filter_data.head()

employee_id         department     region  age  length_of_service  dummy_sum
1          133         Technology  region_15   46                  6         52
3        45914  Sales & Marketing   region_2   36                  9         45
7        13501  Sales & Marketing  region_26   41                  7         48

A new column named `dummy_sum` has been created.

Can we concatenate the string of some columns? Yes we can! We have to use `astype()` to cast a data type.

In [ ]:
filter_data['str_sum'] = filter_data['age'].astype(str) + filter_data['length_of_service'].astype(str)
filter_data.head()

employee_id         department  ... dummy_sum  str_sum
1        22254  Sales & Marketing  ...        47     3611
2        72714         Operations  ...        46      379
6        56336                 HR  ...        55      505
7        28563              Legal  ...        49      436
8        23391         Operations  ...        55      487

[5 rows x 7 columns]

Now, we have another new column named `str_sum` containing concatenated string values of the `age` and the `length_of_service` columns.

# Grouping Dataframe
Pandas has similar method to `GROUP BY` statement in SQL. Grouping means, we will group our data based on specific columns and doing some calculation (aggregation).

## Group by Single Column

### Aggregating All Columns
For example, we want to calculate all columns' average and group it by the `department` column.

In [ ]:
filter_data.groupby('department').mean()

employee_id  ...  is_promoted
department                      ...             
Sales & Marketing      29707.5  ...          0.5
Technology               133.0  ...          0.0

[2 rows x 8 columns]

We have a new dataframe that contains aggregation result.

### Aggregating Specific Columns
What if we only want to know the average age of each department? We can select the `department` and the `age` columns, then group by `department`.

In [ ]:
filter_data[['department','age']].groupby('department').mean()

age
department             
Sales & Marketing  38.5
Technology         46.0

### Multiple Aggregations
The above example only use `mean()` to calculate the average of a column. But, can we use different aggregation method?

We can use `agg()` to do different aggregations.

Syntax


```
dataframe.groupby('column_name').agg(new_col_1 = ('col_name_1', 'agg_method_1'),
                                    (new_col_2 = ('col_name_2', 'agg_method_2'))
```

How to calculate the average age and maximum length of service of each department?


In [ ]:
filter_data.groupby('department').agg(avg_age = ('age', 'mean'),
                                      max_services = ('length_of_service', 'max'))

avg_age  max_services
department                              
Sales & Marketing     38.5             9
Technology            46.0             6

We got a new dataframe with the `department` column becomes the index label. If you don't want to set `department` as index, you can use `reset_index()`.

In [ ]:
filter_data.groupby('department').agg(avg_age = ('age', 'mean'),
                                      max_services = ('length_of_service', 'max')).reset_index()

department  avg_age  max_services
0  Sales & Marketing     38.5             9
1         Technology     46.0             6

## Group by Multiple Columns
Just like sorting, we can group our data using multiple columns.

Let's make a new dataframe containing 100 randomly sampled data.

In [ ]:
data_100 = data.sample(n=100)
data_100.head()

employee_id         department  ... avg_training_score is_promoted
34499        43900              Legal  ...               60.0           0
18084        17913  Sales & Marketing  ...               52.0           0
22621         9850  Sales & Marketing  ...               48.0           0
38969        22908         Technology  ...               78.0           0
42228        58043          Analytics  ...               83.0           0

[5 rows x 13 columns]

I want to know the average age of each department and education.

In [ ]:
data_100[['department','education','age']].groupby(['department','education']).mean().round(2)

age
department        education              
Analytics         Bachelor's        36.14
Finance           Bachelor's        33.33
                  Master's & above  36.00
HR                Bachelor's        35.25
Legal             Bachelor's        32.00
                  Below Secondary   28.00
Operations        Bachelor's        37.46
                  Master's & above  36.40
Procurement       Bachelor's        32.00
                  Master's & above  38.00
R&D               Bachelor's        29.00
Sales & Marketing Bachelor's        36.73
                  Master's & above  37.36
Technology        Bachelor's        34.44
                  Below Secondary   25.00
                  Master's & above  36.80

# Merging Dataframe
Just like SQL JOINs. It used to combine several tables in terms of columns (combining both tables' columns).

Syntax

```
left_table.merge(right_table, how='join_type', on='column_name')
```

To combine 2 tables, we need a connector that will be placed on `on` parameter. In database term, we need a primary key and a foreign key. In this dataset, the `employee_id` acts as the connector.

In [ ]:
data.head()

employee_id         department  ... avg_training_score is_promoted
0        65438  Sales & Marketing  ...               49.0           0
1        65141         Operations  ...               60.0           0
2         7513  Sales & Marketing  ...               50.0           0
3         2542  Sales & Marketing  ...               50.0           0
4        48945         Technology  ...               73.0           0

[5 rows x 13 columns]

I will create 2 dataset from a single dataset
- The first dataset will be the left table. It contains the `employee_id`, `department`, `region`, `education`, `gender`, and `recruitment_channel` columns.

- The second dataset will be the right table. it contains the `employee_id`,	`is_promoted`, and	`no_of_trainings` columns.

In [ ]:
data_left = data.iloc[:20, :6]
data_left.head()

employee_id         department  ... gender recruitment_channel
0        65438  Sales & Marketing  ...      f            sourcing
1        65141         Operations  ...      m               other
2         7513  Sales & Marketing  ...      m            sourcing
3         2542  Sales & Marketing  ...      m               other
4        48945         Technology  ...      m               other

[5 rows x 6 columns]

In [ ]:
data_right = data.loc[18:30, ['employee_id', 'is_promoted', 'no_of_trainings']]
data_right.head()

employee_id  is_promoted  no_of_trainings
18        17436            0                1
19        54461            0                1
20        12067            0                1
21        33332            0                1
22        58789            0                1

## Inner Join
The output is only the intersected data (`B`), the data that appears in both table `A` and table `C`.

<img src='https://raw.githubusercontent.com/adhang/data-science-digitalskola/main/07.%20Intermediate%20Dataframe/res/inner_join.png' width=300>

In [ ]:
data_left.merge(data_right, how='inner', on='employee_id')

employee_id         department  ... is_promoted no_of_trainings
0        17436  Sales & Marketing  ...           0               1
1        54461         Operations  ...           0               1

[2 rows x 8 columns]

There's an intercept between table `A` and table `C` where the `employee_id` is 17436 and 54461. In other words, these two `employee_id` records appear on both tables.

## Left Join
The output is only the intersected data (`B`) and all data that appears on table `A`. The data records on table `A` that do not intersect with/ appear on table `C` will be `NaN`.

<img src='https://raw.githubusercontent.com/adhang/data-science-digitalskola/main/07.%20Intermediate%20Dataframe/res/left_join.png' width=300>

In [ ]:
data_left.merge(data_right, how='left', on='employee_id')

employee_id         department  ... is_promoted no_of_trainings
0         65438  Sales & Marketing  ...         NaN             NaN
1         65141         Operations  ...         NaN             NaN
2          7513  Sales & Marketing  ...         NaN             NaN
3          2542  Sales & Marketing  ...         NaN             NaN
4         48945         Technology  ...         NaN             NaN
5         58896          Analytics  ...         NaN             NaN
6         20379         Operations  ...         NaN             NaN
7         16290         Operations  ...         NaN             NaN
8         73202          Analytics  ...         NaN             NaN
9         28911  Sales & Marketing  ...         NaN             NaN
10        29934         Technology  ...         NaN             NaN
11        49017  Sales & Marketing  ...         NaN             NaN
12        60051  Sales & Marketing  ...         NaN             NaN
13        38401         Technology  ...         NaN             NaN
14        77040                R&D  ...         NaN             NaN
15        43931         Operations  ...         NaN             NaN
16         7152         Technology  ...         NaN             NaN
17         9403  Sales & Marketing  ...         NaN             NaN
18        17436  Sales & Marketing  ...         0.0             1.0
19        54461         Operations  ...         0.0             1.0

[20 rows x 8 columns]

The `is_promoted` and `no_of_trainings` columns are column from table `C`. On these columns, the records from index 0-17 are `NaN` because they only appear on table `A` but not in table `C`. You can see that the last two rows (index 18 and 19) are the same as the result of the inner join.

## Right Join
The output is only the intersected data (`B`) and all data that appears on table `C`. The data records on table `C` that do not intersect with table `A` will be `NaN`.

<img src='https://raw.githubusercontent.com/adhang/data-science-digitalskola/main/07.%20Intermediate%20Dataframe/res/right_join.png' width=300>

In [ ]:
data_left.merge(data_right, how='right', on='employee_id')

employee_id         department  ... is_promoted no_of_trainings
0         17436  Sales & Marketing  ...           0               1
1         54461         Operations  ...           0               1
2         12067                NaN  ...           0               1
3         33332                NaN  ...           0               1
4         58789                NaN  ...           0               1
5         71177                NaN  ...           0               1
6         52057                NaN  ...           0               2
7         26585                NaN  ...           0               1
8         37749                NaN  ...           0               1
9          2370                NaN  ...           0               1
10         8219                NaN  ...           0               1
11        74759                NaN  ...           0               1
12        27395                NaN  ...           0               1

[13 rows x 8 columns]

The `department`, `region`, `education`, `gender`, and `recruitment_channel` columns are column from table `A`. On these columns, the records from index 2-12 are `NaN` because they only appear on table `C` but not in table `A`. You can see that the first two rows (index 0 and 1) are the same as the result of the inner join.

## Outer Join
The output is the combination of all data from table `A` and table `C`. The data records that do not intersect will be `NaN`.

<img src='https://raw.githubusercontent.com/adhang/data-science-digitalskola/main/07.%20Intermediate%20Dataframe/res/outer_join.png' width=300>

In [ ]:
data_left.merge(data_right, how='outer', on='employee_id')

employee_id         department  ... is_promoted no_of_trainings
0         65438  Sales & Marketing  ...         NaN             NaN
1         65141         Operations  ...         NaN             NaN
2          7513  Sales & Marketing  ...         NaN             NaN
3          2542  Sales & Marketing  ...         NaN             NaN
4         48945         Technology  ...         NaN             NaN
5         58896          Analytics  ...         NaN             NaN
6         20379         Operations  ...         NaN             NaN
7         16290         Operations  ...         NaN             NaN
8         73202          Analytics  ...         NaN             NaN
9         28911  Sales & Marketing  ...         NaN             NaN
10        29934         Technology  ...         NaN             NaN
11        49017  Sales & Marketing  ...         NaN             NaN
12        60051  Sales & Marketing  ...         NaN             NaN
13        38401         Technology  ...         NaN             NaN
14        77040                R&D  ...         NaN             NaN
15        43931         Operations  ...         NaN             NaN
16         7152         Technology  ...         NaN             NaN
17         9403  Sales & Marketing  ...         NaN             NaN
18        17436  Sales & Marketing  ...         0.0             1.0
19        54461         Operations  ...         0.0             1.0
20        12067                NaN  ...         0.0             1.0
21        33332                NaN  ...         0.0             1.0
22        58789                NaN  ...         0.0             1.0
23        71177                NaN  ...         0.0             1.0
24        52057                NaN  ...         0.0             2.0
25        26585                NaN  ...         0.0             1.0
26        37749                NaN  ...         0.0             1.0
27         2370                NaN  ...         0.0             1.0
28         8219                NaN  ...         0.0             1.0
29        74759                NaN  ...         0.0             1.0
30        27395                NaN  ...         0.0             1.0

[31 rows x 8 columns]

From this result, we know that the records with index 0-17 are the records that only appear on table `A`, so they have `NaN` values on the `is_promoted` and `no_of_trainings` columns.

On the other hand, the records from index 20-30 are the records that only appear on table `C`, so they have `NaN` values on the `department`, `region`, `education`, `gender`, and `recruitment_channel` columns.

The rows with index 18 and 19 have complete records on all columns, that's because the `employee_id` values exist on both tables.